### ISSUE TITLE
Registrar Backers

### ISSUE THREAD

**Issue by SmithSamuelM** [WebOfTrust/keripy#90]
# Registrar Backers
Unlike witness backers, registrar backers have associated metadata that specifies the registry or ledger upon which the registrar backer anchors the KEL events.

Originally it was proposed that registrar backers be indicated by using a transferable identifier derivation code but have an inception event with a null next field. This would make the registrar identifier effectively non-transferable but provide an inception event that could anchor meta-data. However this would break alot of witness related code which expects that the witness identifiers are explicily not merely effectively non-transferable.

This revised proposal uses explicity non-transferable identifiers for backers but adds a new config trait to indicate that the backers are registrars and adds a new seal type to indicate the anchor for the metadata.

config trait. `RB` for registrar backers or ledger registry backed instead of witness pool backed.

Seal
```python

{


}


```

**Comment by rodolfomiranda** [comment]
Please, confirm if I understand it correctly:
- the `b` label will contain the backer identifier. (can it have other identifiers as well?)
- the `c` label will contain "RB" as a config trait
- the `a` label will contain a Registrar Seal composed by:
-- `bi` label with the backer identifier
-- `d` label with the SAID of the metadata SAD (bare message)

Questions:
1. what is the bare message type? `bar` or `bre`?
2. you said that **attached** to the designating establishment event there's a bare message. What it means that a message is **attached**? what are the mechanism to attach a message?
3. based on [this](https://hackmd.io/@SamuelMSmith/Sy_g84sDK), what's the correct order of the seal labels on a Registrar Seal? `bi`-`d`; or `d`-`bi`
4. is there a definition and how to use the `r`, `rr` and `ri` labels?

**Comment by SmithSamuelM** [comment]
@rodolfomiranda 


I assume you mean the the field labels in an establishment event. 
> the b label will contain the backer identifier. (can it have other identifiers as well?)
In this case no it only has the backer in its list the value of the b field value is a list so its value will be ` [bi]`

> the c label will contain "RB" as a config trait
yes

> the a label will contain a Registrar Seal composed by:

The a field value is a list so one of is entries will b a backer seal

I assume this is refering to a backer seal

> -- bi label with the backer identifier
> -- d label with the SAID of the metadata SAD (bare message)

Yes

Questions:

> what is the bare message type? bar or bre?
`bar`  fixed the inconsistent references above

> you said that attached to the designating establishment event there's a bare message. What it means that a message is attached? what are the mechanism to attach a message?

Because KERI is meant to support any kind of transport protocol and specifically to support streaming on bare metal TCP or eventually UDP, any KERI message type can be interleaved in a stream. So for example if an event message  has the backer seal in it, then in the stream following the event message and any attachments like signatures the following message could be the bare, `bar` message. Its attached to the stream. Alternatively when using http, one could tunnel a stream fragment in the body of the http message. The stream fragment could include both the event message with attachments and also the bare, `bar` message. Althernatively one could have a rest API for the sender of the event message wherein the recipient of and event message with a backer seal could see the seal and then use that as trigger to then request using the prod, `pro' message the corresponding bare, 'bar` message from the sender. All of the above are possible with KERI. 

> based on [this](https://hackmd.io/@SamuelMSmith/Sy_g84sDK), what's the correct order of the seal labels on a Registrar Seal? bi-d; or d-bi

`bi, d`

> is there a definition and how to use the r, rr and ri labels?

The `ri` field in the example has nothing to do with a registrar backer. It was just copied from a different example sorry.
The example has been fixed.

The `r` field is the route. It  is the equivalent of the resource path in a ReST interface but because KERI must support peer-to-peer asynchronous protocols (i.e. can't depend on http or ReST) the messages explicitly include a route to the resource being requested.

The `rr` field is the return route. Because KERI must support peer-to-peer asynchronous protocols (i.e. can't depend on http or ReST) the messages explicitly include a return route so the  the resource being requested is returned to the correct processor on the receiving end. 

The request message, in this case the `pro`, prod, includes both  route, `r` and return route, `rr` fields. The route is the path to the resource on the server host and the return route tells the server host how to return it. The `r` field in the correponding bare, `bar` message is assigned the value of the `rr` field in the triggering `pro`, prod message. An unsolicited bare, `bar` message either leaves the `r` empty or uses a well known value as determined by agreement or the type of data being sent unsolicited. That is application dependent.

**Comment by SmithSamuelM** [comment]
@rodolfomiranda Just to trigger a notice to reread my answers. I substantially updated them since my first response.

**Comment by iFergal** [comment]
This is an old issue but since it's still open I will ask here @SmithSamuelM @rodolfomiranda. :)

In this design, a ledger registrar is designated by its non-transferable AID that also acts as the entry point to the chain (which works for the current [cardano-backer](https://github.com/WebOfTrust/cardano-backer) since Cardano uses Ed25519 keys).

But what happens if a ledger registrar loses access to its keys, so transactions can no longer be pushed on-chain - are all identifiers tied to it not stuck now since a verifier will be looking for ledger transactions from the specific address/AID from the `b` array of 1?

Since we are relying on the consensus algorithm of a ledger itself, rather than witnesses and KA2CE based on `b` and `bt` in the icp/rot event - would it instead be possible to:
- specify the ledger as a config trait - e.g. `CARDANO_MAINNET` (from a known list), with `b = []`,
- push the KEL directly on-chain (no witness, since this isn't KA2CE - but I guess have a validity check),
- have verifiers or watchers (“global watcher”) read all of the chain, and build up an index of the global AID space?
  - this could be narrowed to specific address or pool of addresses too for performance gains based on a specific domain but now the identifiers aren't strongly bound to the entry ledger address.
  - watchers can provide watcher receipts - so verifiers can check with watchers that index the chain.

But maybe I'm missing a specific reason for needing an entry point/backer AID to be marked in the event.

**Non-enterprise advantage:**
In an enterprise setting, having a central ledger registrar that can push events on-chain works. But for an average end user managing their personal identifiers, who may not be able to run their own ledger registrar instance - they would need to find some paid service for this. Since ledger transactions aren't free.

A more decentralised approach is possible if there exists identity wallets that can also operate under the hood as also a crypto wallet for that ledger (optionally). This is something we’re interested in exploring in the [wallet](https://github.com/cardano-foundation/cf-identity-wallet) I work on. If an end user has a funded wallet, they can serialise a transaction with their event themselves and push it to a node in the network, rather than a ledger registrar API. Then a global watcher or verifier can pick it up from the chain.

**Comment by SmithSamuelM** [comment]
To change the registrar as backer identifier, do a rotation event with new registrar identifier.  Any other information associated with that registrar provide in the attached data anchored with the registrar backer seal. A change to the backer identifier is a trigger to look for a new seal in the rotation event.  

We want to avoid putting too much public information in an event. We want the events to be private by default. Putting the registrar information other than the identifier in the sealed information accomplishes that.   All the non-KERI stuff you suggest would likely be Ledger specific so it does not belong in the KERI specification.  The reason for using a  non-transferable identifier for the Registrar is to avoid the regress into having to get a KEL for the registrar.  The registrar identifier does not have to be Ed25519 based it can be any public key based primitive supported by CESR.  So it could be transformation of an Ethereum or Bitcoin address which use ECDSA secp256k1 which is already in the CESR code table.  
So if the Registrar has to move to a new bitcoin address it just issues a rotation event and publishes the new Registrar identifier.

**Comment by SmithSamuelM** [comment]
Alternatively the registrar identifier is not actually a bitcoin address but the identifier of a trusted Oracle for the ledger. The registrate backer sealed data provides any and all details about that.

**Comment by iFergal** [comment]
> To change the registrar as backer identifier, do a rotation event with new registrar identifier. Any other information associated with that registrar provide in the attached data anchored with the registrar backer seal. A change to the backer identifier is a trigger to look for a new seal in the rotation event.

OK, assuming this scenario where the original registrar identifier is completely lost/dead, then the rotation event can only be signed by the new registrar identifier. I guess I was assuming for a rotation event to be valid, it must be signed by the current backer registrar.

The new registrar identifier in this case would also have a new Cardano address. So for discoverability, a verifier needs to know the new address out of band or index all addresses (global watcher - though maybe scoped to a set of AIDs instead of all AIDs).

> Statement made by the secondary root-of-trust need to be verifiable non-repudiable. The AID of the registrar is the trusted source for those statements. They can be signed by that identifier because a non-trans identifier is a public key.

I assume at least that as a verifier/watcher, I should read from the chain instead directly from a registrar backer API (like the current [cardano-backer](https://github.com/WebOfTrust/cardano-backer/commits/main)) because otherwise there's no proof that it actually _went_ on-chain and used the ledger's consensus algorithm. This is sort of like a single normal witness with KA2CE a threshold of 1, if we don't check the chain.

If the secondary root of trust is the ledger itself then I feel like the only proof that it went on the ledger is the valid ledger transaction itself. And not just a non-reputiable signature from the ledger registrar's AID because the registrar could just choose to not put it on chain. (or at least that these are combined, like registrar receipt within a transaction)

**Comment by iFergal** [comment]
Yes sorry when I said signed by the registrar I meant as a receipt signature - the rotation event itself signed by the controller, of course.

If we can always change it, sounds good. It was actually the "joint confirmed witness rotation" section in the white paper for witnesses that made me think it might apply to ledger registrars - but that is also detailed as an optional step for validators. I'll give those sections a re-read.

**Comment by SmithSamuelM** [comment]
Yes so once the rotation event is published by the controller, the old registrar is no longer a registrar, so it no longer has any role to play as a registrar and therefore isn’t needed to sign anything, at least as a registrar. So it won’t be receipting anything.

**Comment by iFergal** [comment]
Thanks, this makes sense!

As a validator or watcher, can you describe what happens if I read transactions from the ledger directly and find the latest KEL but it's not witnessed by the registrar, for whatever reason?

**Comment by SmithSamuelM** [comment]
A ledger backer is essentially replacing a witness pool with the ledger pool of nodes. How that ledger pool "witnesses" the KEL is ledger specific. Its up to whoever decides to use a ledger instead of a witness pool to define what that means. And anyone who wants to interoperate with an AID that has a ledger backer instead of a witness pool of backers needs to understand how that works in order to decide if they want or can interoperate.  So the ledger registrar AID is just a hook to discover the ledger as a secondary root-of-trust. For example, a validator of the ledger as backer could search for transactions on the ledger attributed to the  (public key, address) that is derived from the Registrar AID.  Those transactions for example could anchor the KEL as a whole or the KEL events invididually. Or something like that. That validator would have to accept that they trust the ledger to anchor that KEL and be able to verify those anchors as legitimate and not accept a KEL or a KEL event unless it was anchored. This is similar to not accepting an event on a KEL unless it has beeen witnessed (receipted) by a threshold number of witnesses.  But KERI doen't normatively define what it means to validate a ledger backed KEL. It just provides a mechanism for someone to publish via a KEL that the KEL is ledger backed. 

Each ledger backed mechanism needs to create its own specification that says to anyone who wants to interoperate with it what the validation rules are.  In the early days of KERI, there seemed to be  interest in leveraging existing Ledger infrastructure tooling to provide discovery via the ledger and backing via the ledger. Over time, as ledgers have increasingly come to be seen as platforms (islands) of non-interoperability, the interest in ledger backers has waned.  In fact, to my knowledge, RootsID was the only one to implement a ledger backer for KERI.  We originally thought GLEIF would want Ledger backers for Ethereum and Sovrin but it become clear that there was zero real interest. Most of those advocating for those ledgers actually didn't want KERI at all. 

IMHO the best use of a ledger is not as a secondary root of trust to replace a witness pool but as a tertiary root of trust to provide a permanent anchor of KELs for archival purposes. More like a checkpoint. So a watcher or witness may behind the scenes want to anchor the kels the watch or witness on some ledger (or IPFs) as a reslient permanent archive to back them up.
This hides the complexity of the ledger so that its not a barrier to interoperability.

**Comment by iFergal** [comment]
Thanks for such a detailed explanation; makes perfect sense that it's ledger specific. So it sounds like some of these discussions should be an issue in the cardano-backer repository.

But briefly, I've mostly been curious if we skip using the registrar hook and just post non-receipted KELs (with a seal whose metadata simply specifies the ledger and ledger network, the backer config trait and no/empty backer array) directly on chain from the controller, and a watcher can search the entire ledger space of KEL transactions. Maybe it's just complicating things.

-----

Regarding using ledger registrars: This makes sense too, and I understand the lock in between ledgers. The main reason we've been thinking about this at the Cardano Foundation is to lower the barrier to entry to adopting KERI in our eco-system for end users managing their _personal_ identifiers who can't run their own witnesses, and don't know what witnesses to trust, etc. And our enterprise use cases to use witness pools.

> On the other hand, if you have a watcher who wants to be highly reputable, then you could back that watcher with a ledger as a tertiary root of trust and advertise that the KELs your watcher watches are backed onto a ledger. Now, someone can then use your watcher without having to understand the complexity of the ledger. If others who do understand the complexity of the ledger periodically verify that your watcher's kels are indeed being backed onto the ledger and they publish that, then you get a reputation for your watcher as a more trustable watcher. Likewise for a KERI Witness, or KERI Jury Pool, or KERI Judge.

This idea is super interesting to me. Highly reputable watchers is something I wanted to pursue with this "global watcher" where a validator may query a set of different reputable watchers. Even better if there is a judge too.

But fully in agreement with the DID eco-system - I definitely don't want to also create a silo, so a lot to think about.

**Comment by rodolfomiranda** [comment]
>On the other hand, if you have a watcher who wants to be highly reputable, then you could back that watcher with a ledger as a tertiary root of trust and advertise that the KELs your watcher watches are backed onto a ledger. Now, someone can then use your watcher without having to understand the complexity of the ledger. If others who do understand the complexity of the ledger periodically verify that your watcher's kels are indeed being backed onto the ledger and they publish that, then you get a reputation for your watcher as a more trustable watcher. Likewise for a KERI Witness, or KERI Jury Pool, or KERI Judge.

As @iFergal said, that's really interesting. Realizing that blockchain is the tertiary root of trust, then any KEL can be backed in the ledger independently of having a specific witness to do it, that also  moves  the ledger complexity to the watcher.
